frontend/ui/src/app/subscription/subscription.component.ts

```diff
@@ -0,0 +1,260 @@
+
+  getPreviewPrice(): number {
+    const subscriptionType = this.form.value.subscriptionType;
+    const billingMode = this.form.value.billingMode;
+    const userQty = this.form.value.userAccountQuantity ?? 0;
+    const customerQty = this.form.value.customerOrganizationQuantity ?? 0;
+
+    let userPrice = 0;
+    let customerPrice = 0;
+
+    if (subscriptionType === 'starter') {
+      userPrice = billingMode === 'monthly' ? 19 : 192;
+      customerPrice = billingMode === 'monthly' ? 29 : 288;
+    } else if (subscriptionType === 'pro') {
+      userPrice = billingMode === 'monthly' ? 29 : 288;
+      customerPrice = billingMode === 'monthly' ? 69 : 672;
+    }
+
+    return userPrice * userQty + customerPrice * customerQty;
+  }
+
+  async checkout() {
+    this.form.markAllAsTouched();
+    if (this.form.valid) {
+      try {
+        const body = {
+          subscriptionType: this.form.value.subscriptionType!,
+          billingMode: this.form.value.billingMode!,
+          subscriptionUserAccountQuantity: this.form.value.userAccountQuantity!,
+          subscriptionCustomerOrganizationQuantity: this.form.value.customerOrganizationQuantity!,
+          currency: this.form.value.currency!,
+        };
+
+        // Call the checkout endpoint which will redirect to Stripe
+        await this.subscriptionService.checkout(body);
+      } catch (e) {
+        const msg = getFormDisplayedError(e);
+        if (msg) {
+          this.toast.error(msg);
+        }
+      }
+    }
+  }
+
+  async updateQuantities() {
+    this.form.markAllAsTouched();
+    if (this.form.valid) {
+      try {
+        const body = {
+          subscriptionUserAccountQuantity: this.form.value.userAccountQuantity!,
+          subscriptionCustomerOrganizationQuantity: this.form.value.customerOrganizationQuantity!,
+        };
+
+        // Call the update subscription endpoint
+        const updatedInfo = await this.subscriptionService.updateSubscription(body);
+
+        // Update the subscription info signal with the new data
+        this.subscriptionInfo.set(updatedInfo);
+
+        // Show success message
+        this.toast.success('Subscription updated successfully');
+      } catch (e) {
+        const msg = getFormDisplayedError(e);
+        if (msg) {
+          this.toast.error(msg);
+        }
+      }
+    }
+  }
+
+  getPlanLimits(plan: SubscriptionType): {customers: string; users: string; deployments: string} {
+    const info = this.subscriptionInfo();
+    if (!info) {
+      return {customers: '', users: '', deployments: ''};
+    }
+
+    let limits;
+    switch (plan) {
+      case 'trial':
+        limits = info.trialLimits;
+        break;
+      case 'starter':
+        limits = info.starterLimits;
+        break;
+      case 'pro':
+        limits = info.proLimits;
+        break;
+      case 'enterprise':
+        limits = info.enterpriseLimits;
+        break;
+      default:
+        return {customers: '', users: '', deployments: ''};
```

Comment:
we now have a `never` utility that can be used to ensure switch statements like this are exhaustive, raising a type error if they are not